job-server/src/main/scala/spark/jobserver/BaseSupervisorActor.scala

```diff
@@ -0,0 +1,260 @@
+package spark.jobserver
+
+import scala.collection.mutable
+import scala.util.{Failure, Success, Try}
+
+import akka.actor._
+import akka.cluster.Member
+import akka.cluster.ClusterEvent.MemberUp
+import akka.util.Timeout
+import com.typesafe.config.{Config, ConfigFactory}
+
+import spark.jobserver.common.akka.InstrumentedActor
+import spark.jobserver.JobManagerActor.{SparkContextAlive, SparkContextDead, SparkContextStatus}
+
+/** Messages common to all ContextSupervisors */
+object ContextSupervisor {
+  // Messages/actions
+  case object AddContextsFromConfig // Start up initial contexts
+  case object ListContexts
+  case class AddContext(name: String, contextConfig: Config)
+  case class StartAdHocContext(classPath: String, contextConfig: Config)
+  case class GetContext(name: String) // returns JobManager, JobResultActor
+  case class GetResultActor(name: String)  // returns JobResultActor
+  case class StopContext(name: String, retry: Boolean = true)
+
+  // Errors/Responses
+  case object ContextInitialized
+  case class ContextInitError(t: Throwable)
+  case object ContextAlreadyExists
+  case object NoSuchContext
+  case object ContextStopped
+}
+
+/**
+ * Abstract class that defines all the messages to be implemented
+ * in other supervisors
+ */
+abstract class BaseSupervisorActor extends InstrumentedActor {
+  import ContextSupervisor._
+  import scala.collection.JavaConverters._
+  import scala.concurrent.duration._
+
+  val config = context.system.settings.config
+  val defaultContextConfig = config.getConfig("spark.context-settings")
+
+  import context.dispatcher   // to get ExecutionContext for futures
+  import akka.pattern.ask
+
+  // This is for capturing results for ad-hoc jobs. Otherwise when ad-hoc job dies, resultActor also dies,
+  // and there is no way to retrieve results.
+  val globalResultActor = context.actorOf(Props[JobResultActor], "global-result-actor")
+
+  type SJSContext = (String, Option[ActorRef], Option[ActorRef])
+  // actor name -> (String contextActorName, Option[JobManagerActor] ref, Option[ResultActor] ref)
+  private val contexts = mutable.HashMap.empty[String, SJSContext]
+
+  def wrappedReceive: Receive = {
+    case MemberUp(member) =>
+      onMemberUp(member)
+
+    case ActorIdentity(memberActors, actorRefOpt) =>
+      onActorIdentity(memberActors, actorRefOpt)
+
+    case AddContextsFromConfig =>
+      addContextsFromConfig(config)
+
+    case ListContexts =>
+      sender ! listContexts()
+
+    case AddContext(name, contextConfig) =>
+      val originator = sender()
+      val mergedConfig = contextConfig.withFallback(defaultContextConfig)
+      if (haveContext(name)) {
+        originator ! ContextAlreadyExists
+      } else {
+        registerAndStartContext(name, mergedConfig, false)(
```

Comment:
this is not atomic... an AddContext might be calling `registerAndStartContext` while another call might be executing `haveContext` and register again.